File: java/src/main/java/ram/ka/ru/formats/FlatBuffersSerializer.java
package ram.ka.ru.formats;

import ram.ka.ru.models.User;

import com.google.flatbuffers.FlatBufferBuilder;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.List;


public class FlatBuffersSerializer implements UserSerializer {
    
    public byte[] serialize(User user) throws IOException {
        FlatBufferBuilder builder = new FlatBufferBuilder(1024);
        
        // Создаем строки
        int idOffset = builder.createString(user.getId());
        int nameOffset = builder.createString(user.getName());
        int emailOffset = builder.createString(user.getEmail());
        
        // Создаем массив ролей
        int[] rolesOffsets = new int[user.getRoles().size()];
        for (int i = 0; i < user.getRoles().size(); i++) {
            rolesOffsets[i] = builder.createString(user.getRoles().get(i));
        }
        int rolesVector = ram.ka.ru.models.flatbuffers.User.createRolesVector(builder, rolesOffsets);

        // Создаем объект User
        ram.ka.ru.models.flatbuffers.User.startUser(builder);
        ram.ka.ru.models.flatbuffers.User.addId(builder, idOffset);
        ram.ka.ru.models.flatbuffers.User.addName(builder, nameOffset);
        ram.ka.ru.models.flatbuffers.User.addEmail(builder, emailOffset);
        ram.ka.ru.models.flatbuffers.User.addAge(builder, user.getAge());
        ram.ka.ru.models.flatbuffers.User.addActive(builder, user.isActive());
        ram.ka.ru.models.flatbuffers.User.addRoles(builder, rolesVector);
        ram.ka.ru.models.flatbuffers.User.addBalance(builder, user.getBalance());
        
        int userOffset = ram.ka.ru.models.flatbuffers.User.endUser(builder);
        builder.finish(userOffset);
        
        return builder.sizedByteArray();
    }
    
    public User deserialize(byte[] data) throws IOException {
        ByteBuffer buffer = ByteBuffer.wrap(data);
        ram.ka.ru.models.flatbuffers.User userFlat = ram.ka.ru.models.flatbuffers.User.getRootAsUser(buffer);
        
        User user = new User();
        user.setId(userFlat.id());
        user.setName(userFlat.name());
        user.setEmail(userFlat.email());
        user.setAge(userFlat.age());
        user.setActive(userFlat.active());
        user.setBalance(userFlat.balance());
        
        // Восстанавливаем роли
        for (int i = 0; i < userFlat.rolesLength(); i++) {
            user.getRoles().add(userFlat.roles(i));
        }
        
        return user;
    }
    
    public long measureSerializationTime(User user, int iterations) throws IOException {
        long totalTime = 0;
        for (int i = 0; i < iterations; i++) {
            long startTime = System.nanoTime();
            serialize(user);
            long endTime = System.nanoTime();
            totalTime += (endTime - startTime);
        }
        return totalTime / iterations;
    }
}

File: golang/serializers/flatbuffers_serializer.go
package serializers

import (
	"golang/models"
	"golang/models/flatbuffers"
	"time"

	flatbufferslib "github.com/google/flatbuffers/go"
)

type FlatBuffersSerializer struct{}

func (s *FlatBuffersSerializer) Serialize(user *models.User) ([]byte, error) {
	builder := flatbufferslib.NewBuilder(1024)

	// Создаем строки
	idOffset := builder.CreateString(user.ID)
	nameOffset := builder.CreateString(user.Name)
	emailOffset := builder.CreateString(user.Email)

	// Создаем вектор для ролей
	rolesOffsets := make([]flatbufferslib.UOffsetT, len(user.Roles))
	for i, role := range user.Roles {
		rolesOffsets[i] = builder.CreateString(role)
	}
	flatbuffers.UserStartRolesVector(builder, len(rolesOffsets))
	for i := len(rolesOffsets) - 1; i >= 0; i-- {
		builder.PrependUOffsetT(rolesOffsets[i])
	}
	rolesVector := builder.EndVector(len(rolesOffsets))

	// Создаем объект User
	flatbuffers.UserStart(builder)
	flatbuffers.UserAddId(builder, idOffset)
	flatbuffers.UserAddName(builder, nameOffset)
	flatbuffers.UserAddEmail(builder, emailOffset)
	flatbuffers.UserAddAge(builder, user.Age)
	flatbuffers.UserAddActive(builder, user.Active)
	flatbuffers.UserAddRoles(builder, rolesVector)
	flatbuffers.UserAddBalance(builder, user.Balance)

	userOffset := flatbuffers.UserEnd(builder)
	builder.Finish(userOffset)

	return builder.FinishedBytes(), nil
}

func (s *FlatBuffersSerializer) Deserialize(data []byte) (*models.User, error) {
	userFlat := flatbuffers.GetRootAsUser(data, 0)

	// Восстанавливаем роли
	roles := make([]string, userFlat.RolesLength())
	for i := 0; i < userFlat.RolesLength(); i++ {
		roles[i] = string(userFlat.Roles(i))
	}

	user := &models.User{
		ID:      string(userFlat.Id()),
		Name:    string(userFlat.Name()),
		Email:   string(userFlat.Email()),
		Age:     userFlat.Age(),
		Active:  userFlat.Active(),
		Roles:   roles,
		Balance: userFlat.Balance(),
	}

	return user, nil
}

func (s *FlatBuffersSerializer) MeasureSerializationTime(user *models.User, iterations int) time.Duration {
	var totalTime time.Duration

	for i := 0; i < iterations; i++ {
		start := time.Now()
		s.Serialize(user)
		totalTime += time.Since(start)
	}

	return totalTime / time.Duration(iterations)
}

func (s *FlatBuffersSerializer) MeasureDeserializationTime(data []byte, iterations int) time.Duration {
	var totalTime time.Duration

	for i := 0; i < iterations; i++ {
		start := time.Now()
		s.Deserialize(data)
		totalTime += time.Since(start)
	}

	return totalTime / time.Duration(iterations)
}
